kernel/src/log_compaction/writer.rs

```diff
@@ -0,0 +1,177 @@
+use std::sync::Arc;
+
+use crate::action_reconciliation::RetentionCalculator;
+use crate::log_segment::LogSegment;
+use crate::path::ParsedLogPath;
+use crate::snapshot::Snapshot;
+use crate::table_properties::TableProperties;
+use crate::{DeltaResult, Engine, Error, Version};
+
+use url::Url;
+
+use super::COMPACTION_ACTIONS_SCHEMA;
+use crate::checkpoint::log_replay::{CheckpointBatch, CheckpointLogReplayProcessor};
+use crate::engine_data::FilteredEngineData;
+use crate::log_replay::LogReplayProcessor;
+
+/// Determine if log compaction should be performed based on the commit version and
+/// compaction interval.
+pub fn should_compact(commit_version: Version, compaction_interval: Version) -> bool {
+    // Commits start at 0, so we add one to the commit version to check if we've hit the interval
+    compaction_interval > 0
+        && commit_version > 0
+        && ((commit_version + 1) % compaction_interval) == 0
+}
+
+/// Writer for log compaction files
+///
+/// This writer provides an API for creating log compaction files that aggregate actions
+/// from multiple commit files.
+#[derive(Debug)]
+pub struct LogCompactionWriter {
+    /// Reference to the snapshot of the table being compacted
+    snapshot: Arc<Snapshot>,
+    start_version: Version,
+    end_version: Version,
+    /// Cached compaction file path
+    compaction_path: Url,
+}
+
+impl RetentionCalculator for LogCompactionWriter {
+    fn table_properties(&self) -> &TableProperties {
+        self.snapshot.table_properties()
+    }
+}
+
+impl LogCompactionWriter {
+    pub fn try_new(
```

Comment:
to just expose a narrower API how about making this private and just normalize on the `Snapshot` API for calling this and creating the `LogCompactionWriter`?